sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/PartitionScopeThresholds.java

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.batch;
+
+import com.azure.cosmos.BulkProcessingOptions;
+import com.azure.cosmos.implementation.apachecommons.lang.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.util.function.Tuple2;
+
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class PartitionScopeThresholds<TContext> {
+    private final static Logger logger = LoggerFactory.getLogger(PartitionScopeThresholds.class);
+
+    private final String pkRangeId;
+    private final BulkProcessingOptions<TContext> options;
+    private final AtomicInteger targetMicroBatchSize;
+    private final AtomicLong totalOperationCount;
+    private final AtomicReference<CurrentIntervalThresholds> currentThresholds;
+    private final String identifier = UUID.randomUUID().toString();
+    private final double minRetryRate;
+    private final double maxRetryRate;
+    private final double avgRetryRate;
+
+    public PartitionScopeThresholds(String pkRangeId, BulkProcessingOptions<TContext> options) {
+        checkNotNull(pkRangeId, "expected non-null pkRangeId");
+        checkNotNull(options, "expected non-null options");
+
+        this.pkRangeId = pkRangeId;
+        this.options = options;
+        this.targetMicroBatchSize = new AtomicInteger(options.getMaxMicroBatchSize());
+        this.totalOperationCount = new AtomicLong(0);
+        this.currentThresholds = new AtomicReference<>(new CurrentIntervalThresholds());
+        this.minRetryRate = options.getMinTargetedMicroBatchRetryRate();
+        this.maxRetryRate = options.getMaxTargetedMicroBatchRetryRate();
+        this.avgRetryRate = ((this.maxRetryRate + this.minRetryRate)/2);
+    }
+
+    public String getPartitionKeyRangeId() {
+        return this.pkRangeId;
+    }
+
+    private Pair<Boolean, Boolean> shouldReevaluateThresholds(long totalSnapshot, long currentSnapshot) {
+        if (totalSnapshot < 1_000) {
+
+            return Pair.of(currentSnapshot == 100, false);
+        }
+
+        if (totalSnapshot < 10_000) {
+            return Pair.of(currentSnapshot == 1_000, false);
+        }
+
+        return Pair.of(currentSnapshot % 1_000 == 0, currentSnapshot % 10_000 == 0);
+    }
```

Comment:
So the purpose of this method is to control how fast we are going to revaluate?  and the historical data we are going to keep?

Do we ever reset totalSnapshot? After 10_000 total operation, if the traffic slows down, then reevaluate rate will slow down (revaluate every 1000 operation vs 100 operation)

Reply:
correct - in the beginning reevaluation will happen more quickly.
TotalSnapshot will never be reset. And reevaluation whether we should scale up will happen more quickly (every 1000 operations vs. every 10000 for evaluating whether to scale down)